src/PdfReader/reducer.ts

```diff
@@ -0,0 +1,196 @@
+    }
+
+    /**
+     * Utility function to generate state navigating us to a given resource and page.
+     * Used by multiple cases below.
+     */
+    function goToLocation(index: number, page = 1): PdfState {
+      // only set the resource to null if you're actually changing resources (not just
+      // navigating to a different page in the same resource)
+      const shouldResetResource = state.resourceIndex !== index;
+
+      // check if there is a #startPage in the href of the resource we are navigating to
+      const href = manifest.readingOrder[index].href;
+      const startPage = getStartPageFromHref(href);
+      // only go to the start page if we don't have another valid page we are navigating to
+      // instead.
+      const isNavigatingToEnd = page === -1;
+      const requestedPageIsBeforeStartPage = startPage && page < startPage;
+      const pageNumberToNavigateTo =
+        !isNavigatingToEnd && requestedPageIsBeforeStartPage ? startPage : page;
+
+      const newState = {
+        ...state,
+        resourceIndex: index,
+        pageNumber: pageNumberToNavigateTo,
+      };
+      if (shouldResetResource) {
+        return {
+          ...newState,
+          resource: null,
+          numPages: null,
+        };
+      }
+      return newState;
+    }
+
+    switch (action.type) {
+      case 'ARGS_CHANGED': {
+        return {
+          state: 'ACTIVE',
+          settings: DEFAULT_SETTINGS,
+          resourceIndex: 0,
+          resource: null,
+          pageNumber: 1,
+          numPages: null,
+          scale: 1,
+          pdfWidth: 0,
+          pdfHeight: 0,
+          pageHeight: undefined,
+          pageWidth: undefined,
+          atStart: true,
+          atEnd: false,
+        };
+      }
+
+      case 'GO_FORWARD': {
+        /**
+         * Navigate forward one page or one resource if at the end of the current
+         * resource. Do nothing at the end of the last resource.
+         */
+        // do nothing if we have not parsed the number of pages yet.
+        if (!state.numPages) return state;
+        const atEndOfResource = state.pageNumber === state.numPages;
+        const atEndOfBook =
+          state.resourceIndex === args.manifest.readingOrder.length - 1;
+
+        if (atEndOfResource) {
+          if (atEndOfBook) return state;
+          // go to next resource
+          return goToLocation(state.resourceIndex + 1);
+        }
+        // go to next page
+        return goToLocation(state.resourceIndex, state.pageNumber + 1);
+      }
+
+      case 'GO_BACKWARD': {
+        /**
+         * Navigate backward one page or to the end of the previous resource
+         * if at the beginning of the current resource. Do nothing at the
+         * beginning of the first resource.
+         */
+        // do nothing if we have not parsed the number of pages yet.
+        if (!state.numPages) return state;
+        const atStartOfResource = state.pageNumber === 1;
```

Comment:
This will not work for multi-pdf resources since the start of the resource will be the value from ?startPage=X.